vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,948 @@
+    #[error("Invalid descriptor count {0}")]
+    UnexpectedDescriptorCount(usize),
+    #[error("Invalid descriptor size, expected: {0}, found: {1}")]
+    UnexpectedDescriptorSize(usize, u32),
+    #[error("Tx buf and Rx buf mismatch in descriptor")]
+    TxRxBufOrderMismatch,
+    #[error("Descriptor not found")]
+    DescriptorNotFound,
+    #[error("Descriptor read failed")]
+    DescriptorReadFailed,
+    #[error("Descriptor write failed")]
+    DescriptorWriteFailed,
+    #[error("Failed to send notification")]
+    NotificationFailed,
+    #[error("Failed to create new EventFd")]
+    EventFdFailed,
+}
+
+impl convert::From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+/// The final status written by the device
+const VIRTIO_SPI_TRANS_OK: u8 = 0;
+const VIRTIO_SPI_PARAM_ERR: u8 = 1;
+const VIRTIO_SPI_TRANS_ERR: u8 = 2;
+
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+struct VirtioSpiTransferHead {
+    chip_select_id: u8,
+    bits_per_word: u8,
+    cs_change: u8,
+    tx_nbits: u8,
+    rx_nbits: u8,
+    reserved1: u8,
+    reserved2: u8,
+    reserved3: u8,
+    mode: Le32,
+    freq: Le32,
+    word_delay_ns: Le32,
+    cs_setup_ns: Le32,
+    cs_delay_hold_ns: Le32,
+    cs_change_delay_inactive_ns: Le32,
+}
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioSpiTransferHead {}
+
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+struct VirtioSpiTransferResult {
+    status: u8,
+}
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioSpiTransferResult {}
+
+pub(crate) struct VhostUserSpiBackend<D: SpiDevice> {
+    spi_ctrl: Arc<SpiController<D>>,
+    event_idx: bool,
+    pub exit_event: EventFd,
+    mem: Option<GuestMemoryLoadGuard<GuestMemoryMmap>>,
+}
+
+type SpiDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
+
+impl<D: SpiDevice> VhostUserSpiBackend<D> {
+    pub fn new(spi_ctrl: Arc<SpiController<D>>) -> Result<Self> {
+        Ok(VhostUserSpiBackend {
+            spi_ctrl,
+            event_idx: false,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+        })
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_requests(
+        &self,
+        requests: Vec<SpiDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        let mut reqs: Vec<SpiTransReq> = Vec::new();
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        // Iterate over each SPI request and push it to "reqs" vector.
+        for desc_chain in requests.clone() {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if (descriptors.len() != 3) && (descriptors.len() != 4) {
```

Comment:
We should use the Reader/Writer that @MatiasVara recently merged in virtio-queue, otherwise here we will fail if the driver will change the descriptor layout (see https://github.com/rust-vmm/vhost-device/pull/580).

BTW we can also do later.

Reply:
I think the code would become simpler by using `descriptor_utils`. For example, for the descriptor chain, you instantiate the reader and the writer. Then, if `reader.available_bytes()` > `struct virtio_spi_transfer_req`, you split the reader into two: the header and the `tx_buff`. For the writer, if `writer.available_bytes()` > `struct virtio_spi_transfer_result`, you split the writer into the `tx_buff` and the `result`.

Reply:
Hello @MatiasVara, 
    Here I use 0.12 virtio-queue, but got following error: 
    no method named `iter` found for mutable reference `&mut virtio_queue::queue::Queue` in the current scope

    It happens when parsing requests from vring in the code:
    let requests: Vec<_> = vring
        .get_mut()
        .get_queue_mut()
        .iter(self.mem.as_ref().unwrap().clone())

    
    Currently I have no idea about this, do you have any suggestion? Thank you so much.

Reply:
You have also to update vhost-user-host to 0.15.0. See https://github.com/rust-vmm/vhost-device/pull/580/commits/8a57bfdddd72f0b45167078ff8179ea4e575f352

Reply:
it works, thank you.

Reply:
Resolved